### PR TITLE
[diffusion] feat: TeaCache coefficient calibration

### PR DESCRIPTION
**Pull request by ITerydh**
## Motivation
 
TeaCache is already implemented in sglang, but it only works for models that ship calibrated polynomial `coefficients` and a `teacache_thresh`. There is no tool in the repo to produce them, so any model without hand-copied upstream coefficients simply cannot use TeaCache — LongCat-Image had none, `Wan2_2_T2V_A14B_SamplingParam` shipped no `teacache_params` at all, and MiniMax-H3 rejected `enable_teacache` outright.

This PR adds the missing calibration path, and uses it to enable TeaCache on LongCat-Image (T2I and Edit), Wan2.2-T2V-A14B and MiniMax-H3. On LongCat-Image-Edit this is a ~1.98x denoise speedup that was previously unreachable.

Three things blocked doing this cleanly, all fixed here:                                                                                                                             

1. Calibration needs >=50 diverse prompts to be stable, which is too slow on one GPU, so recording has to be shardable and mergeable across GPUs.
2. MoE dual-transformer models (Wan2.2) need one polynomial per expert; `TeaCacheParams` only had a single shared set.
3. MiniMax-H3 packs video and audio at different timesteps in a single forward, so it has no single-timestep lossless TeaCache contract and simply raised on `enable_teacache`. It needs an opt-in lossy variant rather than the lossless one.

**Docs** — `docs/docs/sglang-diffusion/teacache.mdx`                                                                                                                                 

New "Generating Coefficients (Calibration)" section: what is recorded and fitted, the >=50-diverse-prompt warning, single-GPU and sharded multi-GPU commands, the output format (single vs MoE `high`/`low`), how to wire values into a `SamplingParams` subclass, and how to enable TeaCache at inference. LongCat-Image added to the support table.

> PS: Different models have different coefficients. It is recommended that users generate coefficients for fitting using a diverse test set (preferably with more than 50 examples). It is also recommended to generate corresponding coefficients and thresholds for models with different epochs and resolutions.

## Accuracy Tests

LongCat-Image-Edit, A800, 50 steps, seed 42, `--enable-teacache` at `teacache_thresh=0.13`, compared against the same seed with TeaCache off:

- image mean |delta| = 0.86/255, no visible artifacts.
- TeaCache is opt-in, so default runs are unaffected: with `--enable-teacache` absent, no site is engaged and the output is unchanged.

Wan2.1-14B and Wan2.2-A14B were checked for correct skip behaviour rather than pixel deltas: per-expert coefficient selection and per-CFG-branch skip decisions were verified to pick the expected branch on every step.

The coefficients of MiniMax-H3 are currently generated using only 8 cases, but the results are already satisfactory. If better performance is required, more samples can be added. 

The configuration used is as follows:
all

Unit tests in `python/sglang/multimodal_gen/test/unit/test_teacache_calibrate.py` cover the fit/threshold/merge math on CPU: relative-delta definition, zero-change and single-sample edge cases, flat-run threshold selection including the no-flat-run fallback, single-expert vs MoE output shape, and count-weighted step averaging.

## Speed Tests and Profiling

Denoise time, TeaCache off vs on:

| Model | Off | On | Speedup |
| --- | --- | --- | --- |
| LongCat-Image-Edit (A800, 50 steps, thresh 0.13) | longcat_notc 44.66 s | longcat_tc 22.61 s | ~1.98x |
| Wan2.1-14B (50 step) | wan21_baseline 1323.33 s | wan21_teacache 685.42 s | ~1.93x |
| Wan2.2-T2V-A14B (40 steps, per-expert thresh 0.30) | wan22_baseline 1079.79 s | wan22_teacache 625.08 s | ~1.73x |
| MiniMax-H3 (thresh 0.27) | h3_baseline 148.45s | h3_teacache 68.06s | ~2.18x |

The speed is obtained from my testing environment. The overall environment is fair and comparable.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).                   
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :x: [Run #34451177460](https://github.com/sgl-project/sglang/actions/runs/34451177460)
Latest PR Test (Extra): :x: [Run #34451177251](https://github.com/sgl-project/sglang/actions/runs/34451177251)
Latest PR Test (AMD ROCm 10): :x: [Run #34451177516](https://github.com/sgl-project/sglang/actions/runs/34451177516)

### CODE REVIEW

**Comment by ITerydh** [comment]
@BBuf PTAK ~

File: python/sglang/multimodal_gen/runtime/models/dits/longcat_image.py [hunk 1]
```diff
@@ -903,5 +914,43 @@ def forward(
             return Transformer2DModelOutput(sample=output)
         return output
 
+    def should_skip_forward_for_cached_states(self, **kwargs) -> bool:
+        """TeaCache decision: True to reuse the cached residual this step."""
+        if not self.enable_teacache:
+            return False
+        ctx = self._get_teacache_context()
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/models/dits/longcat_image.py:926]: Could we reset TeaCache explicitly at request start? `_get_teacache_context()` checks the previous CFG branch state, so after serial CFG ends negative, the next request may keep stale residuals.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 2]
```diff
@@ -297,6 +297,10 @@ def __init__(
         super().__init__()
         self.transformer = transformer
         self.transformer_2 = transformer_2
+        # Tag MoE experts (Wan2.2) so TeaCache picks per-expert coefficients.
+        if transformer_2 is not None:
+            transformer._teacache_expert_tag = "high"
+            transformer_2._teacache_expert_tag = "low"
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py:327]: Each expert has a local `cnt`, but skip boundaries use global denoising steps. The low-noise expert also never sees step 0, so its state is not reset. Could we reset both experts per request and use the global step index?

File: python/sglang/multimodal_gen/runtime/cache/teacache_calibrate.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+        expert: str = SINGLE_EXPERT,
+    ) -> None:
+        """Record one transformer call.
+
+        The first call of each (expert, branch) has no predecessor and yields a
+        zero diff, which is harmless since early steps are always computed.
+        """
+        state = self._expert(expert)
+        branch_key = "neg" if is_cfg_negative else "pos"
+        prev = state["prev"].get(branch_key)
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/cache/teacache_calibrate.py:158]: `prev` is never cleared between samples, so each prompt's step 0 is compared with the previous prompt's final step, corrupting the fit. Please clear per-expert/branch history at each sample boundary.

**Review by BBuf — COMMENTED** [review]
A few correctness issues around TeaCache lifecycle and calibration:

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/minimax_h3/stages/denoising.py [hunk 4]
```diff
@@ -700,6 +700,22 @@ def _run_full_loop(self, batch: Req, server_args: ServerArgs) -> None:
...
+            # exposes it; also resets per-step state so calibration counts cleanly).
+            if hasattr(model, "h3_teacache_configure"):
+                tcp = getattr(sampling, "teacache_params", None)
+                num_steps = len(sigmas_video) - 1
+                if tcp is not None:
+                    start, end = tcp.get_skip_boundaries(num_steps, do_cfg=False)
+                else:
+                    start, end = 1, num_steps - 1
+                model.h3_teacache_configure(
+                    enabled=bool(getattr(sampling, "enable_teacache", False)),
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/minimax_h3/stages/denoising.py:774]: This enables the explicitly lossy H3 TeaCache path even under `quality="lossless"`. Please gate it behind `quality="high"` or reject this combination.

File: python/sglang/multimodal_gen/runtime/models/dits/longcat_image.py [hunk 5]
```diff
@@ -874,25 +874,36 @@ def forward(
...
-                temb=temb,
-                image_rotary_emb=image_rotary_emb,
-                cos_sin_cache=cos_sin_cache,
-                positions=positions,
-            )
+        # TeaCache: reuse the cached residual when temb barely changed.
+        if self.should_skip_forward_for_cached_states(temb=temb):
+            hidden_states = self.retrieve_cached_states(hidden_states)
+        else:
+            if self.enable_teacache:
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/models/dits/longcat_image.py:886]: `enable_teacache` defaults to `True`, and LongCat never updates it from the request. TeaCache-off requests still clone hidden states and store residuals. Please gate this work on the request-scoped TeaCache state.

File: python/sglang/multimodal_gen/runtime/cache/teacache.py [hunk 6]
```diff
@@ -297,6 +315,41 @@ def _get_teacache_context(self) -> TeaCacheContext | None:
...
+        calibrator.record(
+            modulated_inp,
+            output,
+            step_index=forward_context.current_timestep,
+            is_cfg_negative=forward_batch.is_cfg_negative,
+            expert=self._teacache_expert_tag,
+        )
+        out_path = envs.SGLANG_TEACACHE_CALIBRATE_OUT
+        if out_path:
+            calibrator.dump_rows(out_path)
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/cache/teacache.py:356]: Every distributed rank can write the same rows file with `"w"`, causing races or truncated JSON. Please use a single-rank/atomic writer or separate per-rank files.

File: python/sglang/multimodal_gen/runtime/models/dits/wanvideo.py [hunk 7]
```diff
@@ -1273,6 +1273,7 @@ def forward(
             # if teacache is enabled, we need to cache the original hidden states
             if self.enable_teacache:
                 self.maybe_cache_states(hidden_states, original_hidden_states)
+            self.maybe_record_calibration(timestep_proj, hidden_states)
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/models/dits/wanvideo.py]: Calibration always records `timestep_proj`, while inference uses `temb` when `use_ret_steps=False`. Please record the same signal selected by the runtime path.

File: python/sglang/multimodal_gen/runtime/cache/teacache_calibrate.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+            current_group.append(int(idx))
+        else:
+            if len(current_group) > len(max_group):
+                max_group = current_group
+            current_group = [int(idx)]
+    if len(current_group) > len(max_group):
+        max_group = current_group
+
+    if max_group:
+        longest_y = y_data[max_group[0] : max_group[-1] + 1]
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/cache/teacache_calibrate.py]: This misses the final point of the flat region. Since slope `i` connects `y[i]` and `y[i+1]`, shouldn't the stop index be `max_group[-1] + 2`?

File: python/sglang/multimodal_gen/runtime/models/dits/longcat_image.py (hunk 1 above)
Comment (ITerydh) [reply python/sglang/multimodal_gen/runtime/models/dits/longcat_image.py:926]: Good catch. Root cause: the reset guard tested self.is_cfg_negative, which still held the previous request's last branch — so a request following one that ended on the negative branch skipped its step-0 reset and could reuse stale residuals. Fixed to use the freshly-read is_cfg_negative from forward_batch, so the reset always fires at request start.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py (hunk 2 above)
Comment (ITerydh) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py:327]: Confirmed — the low-noise expert never sees step 0, so its state carried across requests. Added a per-request reset of both experts. (Global step index left as a follow-up; the reset already bounds it.)

File: python/sglang/multimodal_gen/runtime/cache/teacache_calibrate.py (hunk 3 above)
Comment (ITerydh) [reply python/sglang/multimodal_gen/runtime/cache/teacache_calibrate.py:158]: Confirmed — this corrupted the fit. record now treats step_index == 0 as a sample boundary (prev = None), so step 0 isn't diffed against the previous prompt's last step.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/minimax_h3/stages/denoising.py (hunk 4 above)
Comment (ITerydh) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/minimax_h3/stages/denoising.py:774]: Decoupled instead of gating behind "high" (which is 4×H200-locked). enable_teacache is an explicit lossy opt-in, so it now runs independent of the quality tier and only warns under quality="lossless". I think it is more appropriate to enable it if the user chooses to do so.

File: python/sglang/multimodal_gen/runtime/models/dits/longcat_image.py (hunk 5 above)
Comment (ITerydh) [reply python/sglang/multimodal_gen/runtime/models/dits/longcat_image.py:886]: Confirmed — LongCat never bound enable_teacache to the request. Added self.enable_teacache = forward_batch.enable_teacache at forward start, so teacache-off requests do no clone/cache work.

File: python/sglang/multimodal_gen/runtime/cache/teacache.py (hunk 6 above)
Comment (ITerydh) [reply python/sglang/multimodal_gen/runtime/cache/teacache.py:356]: Confirmed for maybe_record_calibration. Added a global-rank-0 gate before recording/dumping to avoid racing the rows file.

File: python/sglang/multimodal_gen/runtime/models/dits/wanvideo.py (hunk 7 above)
Comment (ITerydh) [reply python/sglang/multimodal_gen/runtime/models/dits/wanvideo.py]: Confirmed. Now records the signal the decision actually diffs — timestep_proj if use_ret_steps else temb — instead of always timestep_proj.

File: python/sglang/multimodal_gen/runtime/cache/teacache_calibrate.py (hunk 8 above)
Comment (ITerydh) [reply python/sglang/multimodal_gen/runtime/cache/teacache_calibrate.py]: Correct off-by-one — slope i spans y[i..i+1], so the flat run ends at max_group[-1]+1. Fixed the slice to +2. Tests pass.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py (hunk 2 above)
Comment (BBuf) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py:327]: The per-request reset fixes state leakage, but it does not fix the boundary semantics. For a 40-step two-expert schedule, the low expert's local `cnt` starts at 0 while `start_skipping`/`end_skipping` remain global (for example 2/39). Its first low steps are therefore classified as early boundaries, and the final global step can still be skipped because the low counter never reaches 39. Please use `ctx.current_timestep` (with the existing CFG semantics) or derive expert-local boundaries, and add a regression test for the expert transition and final step.

File: python/sglang/multimodal_gen/runtime/cache/teacache_calibrate.py (hunk 3 above)
Comment (BBuf) [reply python/sglang/multimodal_gen/runtime/cache/teacache_calibrate.py:158]: `step_index == 0` fixes the single/high expert, but not Wan2.2's low expert: its first call in each prompt occurs at a later global step, so `prev` is still carried from the previous prompt. The denoising-stage reset only clears transformer TeaCache state; it does not clear the process-global calibrator's `_experts["low"]["prev"]`. Please add an explicit per-sample calibrator reset/identity for every expert and a two-prompt dual-expert test.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/minimax_h3/stages/denoising.py (hunk 4 above)
Comment (BBuf) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/minimax_h3/stages/denoising.py:774]: The warning makes the approximation visible, but it does not preserve the API contract: `quality="lossless"` is documented and implemented as the exact/bit-exact reference path. An explicit lossy flag should conflict with an explicit/default lossless tier instead of silently overriding it. Please reject the combination or require a non-lossless quality tier; a warning alone still returns approximate output under the lossless label.

**Review by BBuf — CHANGES_REQUESTED** [review]
### Change comprehension

- Adds a calibration pipeline that records per-step TeaCache input/output deltas, merges prompt shards, fits per-model/per-expert coefficients, and emits reusable parameters.
- Wires calibration and calibrated defaults into Wan, LongCat, and MiniMax-H3, while moving TeaCache state toward request-scoped reset behavior.
- Changes threshold selection and distributed row ownership, and adds unit coverage for the fitting/merge utilities.

```mermaid
flowchart LR
  A[calibration request] --> B[denoising loop]
  B --> C[DiT forward]
  C --> D[record expert / branch deltas]:::changed
  D --> E[rank-0 rows]:::changed
  E --> F[merge prompts]:::changed
  F --> G[fit coefficients + threshold]:::changed
  classDef changed stroke-dasharray: 5 5
```

```mermaid
flowchart LR
  R[new request] --> S[reset transformer TeaCache state]:::changed
  S --> T[select Wan expert / H3 path]
  T --> U[compare step with skip boundaries]:::changed
  U -->|compute| V[refresh residual]
  U -->|skip| W[reuse residual]
  classDef changed stroke-dasharray: 5 5
```

The calibration path is process-global in the worker, but buckets observations by expert and CFG branch. The inference path keeps counters on each transformer instance, including separate Wan2.2 high/low experts.

### Historical review synthesis

I swept all 16 touched paths plus TeaCache/calibration/cache/diffusion/CFG/distributed/benchmark/model keywords across the maintainer corpus: 32,639 inline threads scanned, 89 matching threads across 50 PRs (153 human comments), followed by a conversation-level sweep. The recurring requirements were request-local cache lifecycle (#17140), correct logical-step/CFG accounting (#20706), explicit reset/step boundaries (#19957/#21613/#24227), and accuracy/performance evidence when skip decisions change (#19964/#24227).

### Blocking findings

1. **[P1] Wan2.2 low-expert calibration still crosses sample boundaries.** `TeaCacheCalibrator.record()` clears `prev` only at global `step_index == 0`. The low-noise expert's first call occurs later in the global schedule, so its first observation for prompt N is still diffed against prompt N-1's last low-expert observation. Resetting the transformer state does not reset the process-global calibrator. Please introduce an explicit sample boundary/reset (or sample identity) for every expert and add a two-prompt, dual-expert regression test.

2. **[P1] Wan2.2 still compares expert-local `cnt` with global skip boundaries.** The per-request reset bounds stale state, but `self.cnt < start_skipping or self.cnt >= end_skipping` is still wrong for the low expert. For a 40-step split, its local counter restarts at zero while the boundaries remain global; the first low steps are misclassified and the final global step may not be forced because the low counter never reaches the global end. Please use the global denoising step from the context (including CFG semantics), or convert the boundaries to each expert's local interval, with a regression test covering the expert transition and final step.

3. **[P1] `quality="lossless"` still executes a known-lossy H3 path.** A warning does not preserve the public bit-exact contract established by #33453 and the cookbook. Please reject `enable_teacache=true` with explicit/default lossless, or gate this calibrated path behind a non-lossless quality tier.

The branch also needs the requested rebase reported by `check-changes`. I could run `git diff --check` and compile all changed Python files successfully; the local macOS environment lacks the repository's pytest/torch stack, so the new lifecycle cases above were not executable locally and are not covered by the added tests.

**Review by BBuf — APPROVED** [review]

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py (hunk 2 above)
Comment (ITerydh) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py:327]: Fixed. Added TeaCacheParams.is_step_boundary(current_timestep, num_inference_steps) (per-step, non-CFG-doubled boundaries) and switched wanvideo's should_skip to key on ctx.current_timestep instead of the expert-local cnt. So for a 40-step two-expert schedule the first steps and the final step are forced regardless of which expert runs them; the low expert's early local steps are no longer misclassified as boundaries. Single-transformer behavior is unchanged.

File: python/sglang/multimodal_gen/runtime/cache/teacache_calibrate.py (hunk 3 above)
Comment (ITerydh) [reply python/sglang/multimodal_gen/runtime/cache/teacache_calibrate.py:158]: Fixed. Added TeaCacheCalibrator.begin_sample() which drops prev for every expert, and it's triggered on a global step-0-after-a-nonzero-step (a new sample). This is done in the process-global calibrator itself, not the denoising-stage reset — so the low-noise expert, whose first call is at a later global step, no longer diffs against the previous prompt's final step.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/minimax_h3/stages/denoising.py (hunk 4 above)
Comment (ITerydh) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/minimax_h3/stages/denoising.py:774]: OK — a warning doesn't preserve the contract. Changed to reject: MiniMaxH3SamplingParams._validate now raises when enable_teacache is set with quality="lossless", asking for a non-lossless tier. No more silent lossy output under the lossless label.

**Comment by ITerydh** [comment]
@mickqian @BBuf Could you kindly help me see what I need to do now

**Comment by ITerydh** [comment]
> /tag-and-rerun-ci

PTAL — the remaining CI failures are infrastructure-related: the runner failed to clone lmms-eval during dependency installation, and the other Base failures were fast-fail cascades.

**Comment by ITerydh** [comment]
> > /tag-and-rerun-ci
> 
> PTAL — the remaining CI failures are infrastructure-related: the runner failed to clone lmms-eval during dependency installation, and the other Base failures were fast-fail cascades.

@mickqian @BBuf

**Review by mickqian — CHANGES_REQUESTED** [review]
Current-head CI (42271a12196988293f5c273d473e1335beb30549) still has deterministic TeaCache quality regressions, so another rerun will not help:

- `wan2_1_t2v_1.3b_teacache_enabled`: sim 0.9680 < 0.97, SSIM 0.8900 < 0.92, PSNR 28.5566 < 30.0. These values are identical to the previous run.
- `wan2_2_t2v_a14b_teacache_2gpu`: sim 0.9439, SSIM 0.4373, PSNR 14.5506 versus thresholds 0.97 / 0.90 / 29.1.

Please fix or recalibrate the shipped Wan coefficients and rerun the focused TeaCache cases. I am not treating the 8 ms LTX latency overage or the small non-TeaCache Wan runtime-VRAM overage as independent code failures.

**Comment by ITerydh** [comment]
> Current-head CI ([42271a1](https://github.com/sgl-project/sglang/commit/42271a12196988293f5c273d473e1335beb30549)) still has deterministic TeaCache quality regressions, so another rerun will not help:
> 
> * `wan2_1_t2v_1.3b_teacache_enabled`: sim 0.9680 < 0.97, SSIM 0.8900 < 0.92, PSNR 28.5566 < 30.0. These values are identical to the previous run.
> * `wan2_2_t2v_a14b_teacache_2gpu`: sim 0.9439, SSIM 0.4373, PSNR 14.5506 versus thresholds 0.97 / 0.90 / 29.1.
> 
> Please fix or recalibrate the shipped Wan coefficients and rerun the focused TeaCache cases. I am not treating the 8 ms LTX latency overage or the small non-TeaCache Wan runtime-VRAM overage as independent code failures.请修复或重新校准已发送的 Wan 系数，然后重新运行那些针对 TeaCache 情况的测试。我不会将 8 毫秒的 LTX 延迟超标情况，以及小型非 TeaCache 场景中的运行时间-显存超标情况，视为独立的代码故障。

@mickqian Hi.  Thanks for the detailed CI numbers. A few clarifications:

 Wan2.1 — We did not touch the shipped Wan2.1 coefficients (_wan_1_3b_coefficients / _wan_14b_coefficients); git confirms their last change is the upstream unify refactor #20706, not this PR. Our runtime boundary change is provably equivalent for single-transformer Wan (start=5/end=1.0 forces the same first-N/last steps whether keyed on the doubled per-forward cnt or the global step), so the 1.3B numbers are not a regression introduced here. 

 Wan2.2 — Before this PR, Wan2.2 shipped with teacache_params=None, i.e. enable_teacache was silently a no-op (output identical to baseline). So the wan2_2_t2v_a14b_teacache_2gpu case was effectively comparing baseline-to-baseline. This PR is the first time TeaCache actually runs for Wan2.2, so a nonzero difference vs the baseline is expected and correct behavior, not a new defect — it's the difference between "cache disabled" and "cache enabled."

 On the metrics themselves — sim / SSIM / PSNR against the non-cached reference measure divergence from a specific sample, not quality. For a generative diffusion model, a small perturbation to the denoising trajectory (which is exactly what a lossy cache like TeaCache introduces — lossy means "different parameters/trajectory," not "worse") can move the sample to an equally-valid neighbor: e.g. prompt: "a cat" :  walks vs. lies down are both correct, yet score very low SSIM/PSNR against each other. So these numbers are a rough smoke-test at best and don't quantify visual degradation; TeaCache is explicitly a lossy accelerator and will always shift them. I reviewed the actual Wan2.2 TeaCache output and it's a clean, coherent video (no artifacts/noise) — the acceleration is working as intended. 

 One more point: metric gates like sim/SSIM/PSNR-against-reference are appropriate for lossless optimizations (fused kernels, layout changes, math-equivalent rewrites), where any divergence genuinely signals a bug. They are not the right acceptance criterion for cache-based accelerators like CacheDiT or TeaCache, which are lossy by design and intentionally alter the denoising trajectory — so a lower similarity score is the expected outcome, not a failure signal. For these paths, correctness should be judged by output coherence/quality (no artifacts, prompt adherence), not by bit-level similarity to the uncached reference.

**Comment by ITerydh** [comment]
> Current-head CI ([42271a1](https://github.com/sgl-project/sglang/commit/42271a12196988293f5c273d473e1335beb30549)) still has deterministic TeaCache quality regressions, so another rerun will not help:
> 
> * `wan2_1_t2v_1.3b_teacache_enabled`: sim 0.9680 < 0.97, SSIM 0.8900 < 0.92, PSNR 28.5566 < 30.0. These values are identical to the previous run.
> * `wan2_2_t2v_a14b_teacache_2gpu`: sim 0.9439, SSIM 0.4373, PSNR 14.5506 versus thresholds 0.97 / 0.90 / 29.1.
> 
> Please fix or recalibrate the shipped Wan coefficients and rerun the focused TeaCache cases. I am not treating the 8 ms LTX latency overage or the small non-TeaCache Wan runtime-VRAM overage as independent code failures.

@mickqian @BBuf  Please take a look at the reply. I really don't know how to proceed

**Comment by BBuf** [comment]
> > Current-head CI ([42271a1](https://github.com/sgl-project/sglang/commit/42271a12196988293f5c273d473e1335beb30549)) still has deterministic TeaCache quality regressions, so another rerun will not help:
> > 
> > * `wan2_1_t2v_1.3b_teacache_enabled`: sim 0.9680 < 0.97, SSIM 0.8900 < 0.92, PSNR 28.5566 < 30.0. These values are identical to the previous run.
> > * `wan2_2_t2v_a14b_teacache_2gpu`: sim 0.9439, SSIM 0.4373, PSNR 14.5506 versus thresholds 0.97 / 0.90 / 29.1.
> > 
> > Please fix or recalibrate the shipped Wan coefficients and rerun the focused TeaCache cases. I am not treating the 8 ms LTX latency overage or the small non-TeaCache Wan runtime-VRAM overage as independent code failures.
> 
> @mickqian @BBuf Please take a look at the reply. I really don't know how to proceed

I will track this ci and merged it once the diffusion ci passed, don't worry.
